src/librustc_codegen_llvm/attributes.rs

```diff
@@ -267,31 +266,24 @@ pub fn from_fn_attrs(
         // In panic=abort mode we assume nothing can unwind anywhere, so
         // optimize based on this!
         false
-    } else if codegen_fn_attrs.flags.contains(CodegenFnAttrFlags::UNWIND) {
-        // If a specific #[unwind] attribute is present, use that
-        true
     } else if codegen_fn_attrs.flags.contains(CodegenFnAttrFlags::RUSTC_ALLOCATOR_NOUNWIND) {
-        // Special attribute for allocator functions, which can't unwind
+        // Special attribute for allocator functions, which can't unwind.
         false
     } else if let Some(id) = id {
         let sig = cx.tcx.normalize_erasing_late_bound_regions(ty::ParamEnv::reveal_all(), &sig);
         if cx.tcx.is_foreign_item(id) {
-            // Foreign items like `extern "C" { fn foo(); }` are assumed not to
-            // unwind
+            // Foreign items like `extern "C" { fn foo(); }` and `extern "Rust" { fn bar(); }`
+            // are assumed not to unwind.
```

Comment:
Do we really want to assume that an imported `extern "Rust" { fn bar(); }` does not unwind? I just expanded the comment here to describe already existing behavior.

Cc @gnzlbg

Reply:
> Do we really want to assume that an imported extern "Rust" { fn bar(); } does not unwind?

IIRC, `"Rust"`, `"platform-intrinsic"`, `"rust-intrinsic"`, and `"rustcall"` are actually the same ABI, and it is ok for extern declarations of those to actually unwind. One such a declaration would be the integer `add` intrinsic that panics on integer overflow. We also sometimes import `llvm` intrinsics (via `#[feature(link_llvm_intrinsics(]`) using the `"unadjusted"` ABI. I'm not sure if that one should panic as well.

Reply:
My personal inclination is to just remove the `foreign_item` case here. But that decision should probably be made by someone who actually knows the surrounding code.

Reply:
In fact I had to remove that branch to make the codegen tests work as expected. Whatever that means.